uint/benches/bigint.rs

```diff
@@ -34,383 +33,574 @@ impl U256 {
 	}
 }
 
-use test::{Bencher, black_box};
-
-#[bench]
-fn u256_add(b: &mut Bencher) {
-    b.iter(|| {
-        let n = black_box(10000);
-        let zero = black_box(U256::zero());
-        (0..n).fold(zero, |old, new| {
-            old.overflowing_add(U256::from(black_box(new))).0
-        })
-    });
-}
-
-#[bench]
-fn u256_sub(b: &mut Bencher) {
-    b.iter(|| {
-        let n = black_box(10000);
-        let max = black_box(U256::max_value());
-        (0..n).fold(max, |old, new| {
-            old.overflowing_sub(U256::from(black_box(new))).0
-        })
-    });
-}
-
-#[bench]
-fn u512_sub(b: &mut Bencher) {
-    b.iter(|| {
-        let n = black_box(10000);
-        let max = black_box(U512::max_value());
-        (0..n).fold(max, |old, new| {
-            let new = black_box(new);
-            let p = new % 2;
-            old.overflowing_sub(U512([p, p, p, p, p, p, p, new])).0
-        })
-    });
-}
-
-#[bench]
-fn u512_add(b: &mut Bencher) {
-    b.iter(|| {
-        let n = black_box(10000);
-        let zero = black_box(U512::zero());
-        (0..n).fold(zero, |old, new| {
-            let new = black_box(new);
-            old.overflowing_add(U512([new, new, new, new, new, new, new, new]))
-                .0
-        })
-    });
-}
-
-#[bench]
-fn u512_mul(b: &mut Bencher) {
-    b.iter(|| {
-        (1..10000).fold(black_box(U512::one()), |old, new| {
-            old.overflowing_mul(U512::from(black_box(new | 1))).0
-        })
-    });
-}
-
-#[bench]
-fn u512_mul_small(b: &mut Bencher) {
-    b.iter(|| {
-        (1..153).fold(black_box(U512::one()), |old, _| {
-            old.overflowing_mul(U512::from(black_box(10))).0
-        })
-    });
-}
-
-#[bench]
-fn u512_rem(b: &mut Bencher) {
-	let one = U512([8326634216714383706, 15837136097609390493, 13004317189126203332, 7031796866963419685, 12767554894655550452, 16333049135534778834, 140317443000293558, 598963]);
-	let two = U512([11707750893627518758, 17679501210898117940, 2472932874039724966, 11177683849610900539, 2096410819092764509, 8483673822214032535, 36306297304129857, 3453]);
-    b.iter(|| {
-        black_box(one % two);
-    });
-}
-
-#[bench]
-fn u256_rem(b: &mut Bencher) {
+use criterion::{black_box, Bencher, Criterion};
+use std::str::FromStr;
+
+criterion_group!(
+	bigint,
+	u256_add,
+	u256_sub,
+	u256_mul,
+	u256_mul_small,
+	u256_mul_full,
+	u256_div,
+	u256_rem,
+	u256_rem_small,
+	u256_bit_and,
+	u256_bit_or,
+	u256_bit_xor,
+	u256_not,
+	u256_ord,
+	u256_shl,
+	u256_shr,
+	u256_from_le,
+	u256_from_be,
+	u512_add,
+	u512_sub,
+	u512_mul,
+	u512_mul_small,
+	u512_div,
+	u512_rem,
+	u512_rem_small,
+	u512_bit_and,
+	u512_bit_or,
+	u512_bit_xor,
+	u512_not,
+	u512_ord,
+	u512_shl,
+	u512_shr,
+	u128_mul,
+	from_fixed_array,
+);
+criterion_main!(bigint);
+
+fn u256_add(b: &mut Criterion) {
+	b.bench_function("u256_add", |b| bench_u256_add(b));
```

Comment:
Why did you decide to go this route, with the benchmark separate from the fn? Did you notice a difference from, e.g.

```rust
fn benc_name(b: &mut Criterion) {
  b.bench_function("name", |b| {
     // setup
     b.iter(|| {
        // code under test
      })
   }
}
```

The [general advice in the docs](https://bheisler.github.io/criterion.rs/book/faq.html#how-should-i-benchmark-small-functions) is to avoid benchmarking in a loop like this and instead just run the code, `overflowing_add` in this case. Each sample is multiple iterations anyway?

Reply:
> Why did you decide to go this route, with the benchmark separate from the fn?

Because I wrote a quick and dirty emacs macro to convert the benchmarks to Criterion :) 
Can change that ofc. 
> claim it's pretty good at inserting black_box where needed
> to avoid benchmarking in a loop like this

I haven't touched the existing tests logic, but my guess is e.g. in `u256_add` the point of the `fold` has more to do with adding different numbers (not just some random constants), rather than the loop itself.

Reply:
> adding different numbers (not just some random constants)

Fair point. Still, it'd be nice to check if such "tricks" are still needed with Criterion. Ideally we can avoid them and it'd be good to establish what "best practice" is as more and more benchmarks get converted over.

Reply:
> go this route, with the benchmark separate from the fn?

Pushed a change to fix this 
> claim it's pretty good at inserting black_box where needed

actually, it did make a difference, at least for the last expression